app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/SubmitMessageController.scala

```diff
@@ -30,32 +39,80 @@ import scala.xml.NodeSeq
 class SubmitMessageController @Inject()(
                                          authAction: AuthAction,
                                          xmlParser: ParseXmlAction,
-                                         validateErnInMessageAction: ValidateErnInMessageAction,
-                                         validateMovementIdAction: ValidateMovementIdAction,
                                          submissionMessageService: SubmissionMessageService,
+                                         movementService: MovementService,
                                          workItemService: WorkItemService,
-                                         ernsMapper: ErnsMapper,
+                                         messageValidator: MessageValidation,
+                                         dateTimeService: DateTimeService,
                                          cc: ControllerComponents
                                        )(implicit ec: ExecutionContext)
   extends BackendController(cc) {
 
   def submit(movementId: String): Action[NodeSeq] = {
 
-    (authAction
-      andThen xmlParser
-      andThen validateErnInMessageAction
-      andThen validateMovementIdAction(movementId)).async(parse.xml) {
+    (authAction andThen xmlParser).async(parse.xml) {
       implicit request =>
-        val ern = ernsMapper.getSingleErnFromMessage(request.message, request.validErns)
 
-        workItemService.addWorkItemForErn(ern, fastMode = true)
-
-        submissionMessageService.submit(request).flatMap {
-          case Right(_) => Future.successful(Accepted(""))
-          case Left(error) => Future.successful(error)
+        val result = for {
+          movement <- validateMovementId(movementId)
+          authorisedErn <- validateMessage(movement, request.ieMessage, request.erns)
+          _ <- sendRequest(request, authorisedErn)
+        } yield {
+          Accepted
         }
 
+        result.merge
+    }
+
+  }
+
+  private def validateMovementId(movementId: String): EitherT[Future, Result, Movement] = {
```

Comment:
Can we re-use this for getMessageController too? We could have a MovementIdValidator or similar to be reused in both places

Reply:
done